lib/spack/spack/compilers/clang.py

```diff
@@ -208,6 +208,11 @@ def extract_version_from_output(cls, output):
             r'^Apple (?:LLVM|clang) version ([^ )]+)|'
             # Normal clang compiler versions are left as-is
             r'clang version ([^ )]+)-svn[~.\w\d-]*|'
+            # Starting with clang-9, Ubuntu no longer uses the 'svn' prefix
+            r'clang version ([^ )]+?)-[~.\w\d-]*|'
+            # The order here is important:
+            #        The greedy capture (below) will match before
+            #        the non-greedy one (above)
```

Comment:
I don't fully understand the meaning of this comment. Matching order here is determined by the `|` and by the rule:

> As the target string is scanned, REs separated by '|' are tried from left to right. When one pattern completely matches, that branch is accepted. 

see [Python's regex docs](https://docs.python.org/3/library/re.html). Can you please clarify what you mean?

Reply:
I agree that's confusing. I wasn't referring to the behavior of `re.match`. At the time, I couldn't prove to myself that the non-greedy version would not match a string that the greedy version would- thus creating the possibility of having an input where the ordering of the regexes would determine the correctness of the result. However, this is now moot as I've found a version string that is not correctly parsed by the non-greedy version, but *is* correctly parsed by the greedy one.

    $ lsb_release -d && clang --version
	Description:	Ubuntu 19.04
	clang version 8.0.0-3 (tags/RELEASE_800/final)
	Target: aarch64-unknown-linux-gnu
	Thread model: posix
	InstalledDir: /usr/bin

This produces a version string of '8.0.0' instead of '8.0.0-3'. Does Spack allow compilers with a hyphen in there name as a spec (i.e., '%clang@8.0.0-3')?

Reply:
Yes, versions are allowed to contain periods, hyphens, underscores, and alphanumeric characters.

Reply:
Can you add that output as another test?

Reply:
Honestly I would be fine with Spack parsing the above as `8.0.0`. Everything after the patch level identifies somehow additions to the vanilla `8.0.0` (either pre-release or build information) assuming the Ubuntu packages follow [semver](https://semver.org/). Parsing it as `8.0.0-3` would make sense if we foresee cases where people have multiple `clang` installed all based on the `8.0.0` version which probably is an extremely rare occurrence. @adamjstewart What do you think?

Reply:
@hainest Can you add the output at https://github.com/spack/spack/pull/14365#discussion_r363063974 as a further test and make sure it parse as `8.0.0` (given the thumbs up from @adamjstewart at https://github.com/spack/spack/pull/14365#discussion_r363084325) ? Feel free to ping me back when you're done!

Reply:
@alalazo @adamjstewart Reduced to just the non-greedy match, and added two tests with hyphens in the version. Tests pass for me locally.